ARCs/arc-draft_alternative_0014.md

```diff
@@ -0,0 +1,283 @@
+- `Algorand address` (PKa) to be authenticated;
+- `nonce` representing a unique/random value generated by the verifier.
+
+The JSON structure for such an object is:
+
+```typescript
+interface AuthMessage {
+ /** The domain name of the service */
+ service: string;
+ /** Optional, description of the service */
+ desc?: string;
+ /** Algorand account to authenticate with*/
+ authAcc: string;
+ /** Challenge generated server-side */
+ nonce: string;
+}
+```
+
+For example:
+
+```json
+{
+ "service": "www.servicedomain.com",
+ "desc": "Domain offers important services to users",
+ "authAcc": "KTGP47G64KCXWJS64W7SGJNKTHE37TYDCI64USXI3XOYE6ZSH4LCI7NIDA",
+ "nonce": "1234abcde!%",
+}
+```
+
+The `nonce` field **MUST** be unique for each authentication and **MUST NOT** be used more than once to avoid replay attacks.
+
+Users sign *Authentication Message*s using the secret keys stored into a wallet.
+
+***Problem***: Most of the Algorand wallets can only sign Algorand `Transaction` objects, and do not offer the possibility to sign arbitrary bytes. To overcome such a limitation, the ARC-0014 distinguishes two authentication objects, namely the **Simple Authentication Message**, and the **Transaction Authentication Message**. The former **SHOULD** be used by any signing mechanisms able to sign random bytes, whereas the latter can be used with almost any Algorand wallet.
+
+#### Simple Authentication Message
+
+The *Simple Authentication Message* is a sequence of bytes representing an ARC-0014 *Authentication Message*. It **SHOULD** be used in contexts that support signatures of random bytes. A *Simple Authentication Message* is an *Authentication Message* prepended with the prefix `arc14` for domain separation. It **MUST** be represented as the hash SHA-512/256 of such prefix, together with a msgpack encoded *Authentication Message*. For example, given an *Authentication Message* object `aut_message`, its *Simple Authentication Message* representation would be something like: `SHA512_256(“arc14”+msgpacked_auth_message)`.
+
+#### Transaction Authentication Message
+
+A *Transaction Authentication Message* is an *Authentication Message* represented as an Algorand `Transaction` object. Such a transaction must have the following characteristics:
+
+- it **SHOULD** be an Algorand `Payment Transaction`;
+- it **MUST** include a *Simple Authentication Message* into the transaction's `note` field;
+- it **MUST** be *invalidated* i.e. not executable on any official Algorand network.
+
+> Malicious users could intercept and execute it causing unexpected consequences. For instance, in case of a **MainNet** transaction, it could be intercepted by a malicious user and executed, burning some Algos from the sender’s account due to the payment of fees.
+
+The fields of a `Payment Transaction` object which represents a *Transaction Authentication Message* **MUST** be initialized as follows
+
+- `amount` = 0;
+- `sender/receiver` = *PKa*;
+- `firstValid/lastValid` = 0;
```

Comment:
In my original ARC0014 standard the first valid and last valid is used to ensure time session timeout. Please put it back. Thanks

Reply:
With ARC14 we are not proposinf a standard to handle sessions, but just a way to use Algorand accounts as authentication mean. First/last valid are set to 0 just to invalidate the transaction.

Reply:
as long as wallets have no trouble signing the fv/lv 0, I think this should say and put application specific stuff like timeout in the note field

Reply:
> With ARC14 we are not proposing a standard to handle sessions, but just a way to use Algorand accounts as authentication mean. First/last valid are 0 just to invalidate the transaction (we don't want to execute that transaction on-chain).

Then remove all session related material from the ARC.

Also your comment above that typically a username/password authentication would typically result in an access token might be true for simple line of business apps hosted on private network, but in my experience developing commercial or publicly available systems this is nearly never the case.  But more importantly an access token is not a session token, and the idea of session is quite a loaded term. 

If we're just talking about authenticating at an Identity Provider and offering a method of doing so by signing a message of a specific format, then let's just leave it at that. Beyond that you're going to have to spec out all kinds of auth flows and scenarios.

Reply:
guys, it is very crucial this is there.. i do not agree the timeout of the session should be in the note field..

inability to submit to the blockchain is done through setting the fee to zero.. right?

Reply:
We might want to split this into 2 ARCs, one for the connection & one for the session's handling. Those are two different subjects. 
However, I would like to know why you think @scholtz that using the note field is not optimal?

Reply:
A zero fee does not prevent submitting to blockchain, as pooled fees would allow a different transaction to pay the fee.

Reply:
if someone decides to force store that auth message to blockchain it is his design choice.. he can always make it online through ftp or some other hosting.. zero fee is to prevent user to submit it to blockchain directly

session integrity must be resolved at this standard for sake of security. not field in my original was used as realm.. the message initiated by server so that one person dont take the auth message from one service and use it in other.

the note field should be readable by user so that he can verify what he is signing.. if you guys create complicated structure within the note field users will never check it and you create security leak 

therefore i belive in the note field should be minimum information and we can use other inputs to check for the validity of the message such as the params object .. the net you are submitting to, the rounds validity and so on..

all of this is already implemented in open source applications.. there is AlgorandAuthentication nuget package, there is hasura graphql authenticator, there are few frontend implementations. does anybody of you is already using this standard? whatever you change here i will have to modify existing implementations

Reply:
ARC14 must focuss on authentication. We will leave the session stuff for another ARC. I will now remove the session guidelines from ARC14.

Reagarding the notes field: I see your point and you are probably right. If we obfuscate the note, it will be impossible to verify what message users are signing (making the protocolo vulnerable to MITM).